### DISCUSSION TITLE
Performance improve in get the object

### DISCUSSION THREAD

**Discussion by hxnan** [minio/minio#18780] [Q&A] [answered]
## I think the code below is not work efficiently.

```java
func (z *erasureServerPools) getLatestObjectInfoWithIdx(ctx context.Context, bucket, object string, opts ObjectOptions) (ObjectInfo, int, error) {
	object = encodeDirObject(object)
	results := make([]struct {
		zIdx int
		oi   ObjectInfo
		err  error
	}, len(z.serverPools))
	var wg sync.WaitGroup
	for i, pool := range z.serverPools {
		wg.Add(1)
		go func(i int, pool *erasureSets) {
			defer wg.Done()
			results[i].zIdx = i
			results[i].oi, results[i].err = pool.GetObjectInfo(ctx, bucket, object, opts)
		}(i, pool)
	}
	wg.Wait()

```

```java
func (er erasureObjects) getObjectFileInfo(ctx context.Context, bucket, object string, opts ObjectOptions, readData bool) (FileInfo, []FileInfo, []StorageAPI, error) {
	var mu sync.Mutex
......
	// Ask for all disks first;
	go func() {
		ctx, cancel := context.WithCancel(ctx)
		defer cancel()


```

## Describe the solution you'd like
Add a metaService or cache to remember where is the object( which server and which disk has the object), instead of ask all server and disk.

**Comment by klauspost** [comment, marked as answer]
How would you invalidate this? You cannot just cache local entries and hope that no-one else wrote to it.

MinIO is "read-after-write" consistent so we cannot rely on a "right 90% of the time" approach.

#18239 adds an experimental top-level cache, but it isn't really a good solution.

**Comment by hxnan** [comment]
This about this: If we have cluster with  a lot of server nodes.  A s3 file object (id or other info) should tell the sever:  where i am in 90% time,  if not found then, try other or  all  server.
Maybe the s3Object id can contain the server info, while the object has not be move between the servers, the get and search work will be easyly.

**Reply by klauspost** [reply]
We already use deterministic hashing on object names for  placement.
